frontend/src/GameRouter.tsx

```diff
@@ -1,29 +1,54 @@
-import React from "react";
-import {
-  MemoryRouter,
-  Switch,
-  Route,
-  Redirect,
-  useParams,
-} from "react-router-dom";
+import React, { useContext, useState } from "react";
+import { Router, Switch, Route, Redirect, useParams } from "react-router-dom";
+import { createMemoryHistory } from "history";
 import { BasicPage, NicknamePage, LobbyPage } from "./pages";
+import socket from "./socket";
+import {
+  Player,
+  PlayersContext,
+} from "./ContextProviders/PlayersContextProvider";
+
+export type Settings = {
+  roundLimit?: number;
+  maxPlayers?: number;
+};
 
 type PathParams = {
   gameCode: string;
 };
 
+const memoryHistory = createMemoryHistory();
+
 const GameRouter = () => {
   const { gameCode } = useParams<PathParams>();
+  const { setHost, initialisePlayers, addPlayer, removePlayer } = useContext(
+    PlayersContext
+  );
+  const [settings, setSettings] = useState<Settings>();
+
+  socket.on("navigate", (route: string) => memoryHistory.push(route));
+  socket.on("host", (host: string) => setHost(host));
+  socket.on("players:initial", (players: Player[]) =>
+    initialisePlayers(players)
+  );
+  socket.on("players:add", (nickname: string) => addPlayer(nickname));
+  socket.on("players:remove", (nickname: string) => removePlayer(nickname));
+  socket.on("settings:initial", (initialSettings: Settings) =>
+    setSettings(initialSettings)
+  );
+  socket.on("settings:update", (newSettings: Settings) =>
+    setSettings(newSettings)
+  );
```

Comment:
To reduce network traffic only the setting being updated is sent to the client. I recommend using something similar to the following:

```
socket.on("settings:update", (setting, value) => {
    let key;
    switch (setting) {
      case "MAX_PLAYERS":
        key = "maxPlayers";
        break;
      case "ROUND_LIMIT":
        key = "roundLimit";
        break;
      default:
        key = null;
    }
    if (key !== null) {
      const newSettings = {
        ...settings,
        [key]: value,
      };
      setSettings(newSettings);
    }
  });
```

Reply:
I could also look into having the server return `setting` as the appropriate key to remove this case if you prefer